### PR TITLE
Bug 1804377 - Refactor TabHistory using compose.

### PR DESCRIPTION
**Pull request by iorgamgabriel**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1804377

### CODE REVIEW

File: android-components/plugins/dependencies/src/main/java/DependenciesPlugin.kt [hunk 1]
```diff
@@ -53,7 +53,7 @@ object Versions {
 
     // see https://android-developers.googleblog.com/2022/06/independent-versioning-of-Jetpack-Compose-libraries.html
     // for Jetpack Compose libraries versioning
-    const val compose_version = "1.4.3"
+    const val compose_version = "1.5.0-alpha01"
```
Comment (gabrielluong) [inline android-components/plugins/dependencies/src/main/java/DependenciesPlugin.kt:56]: Please do not ship alpha/beta versions of dependencies.

File: fenix/app/src/main/java/org/mozilla/fenix/tabhistory/TabHistoryBottomSheetView.kt [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+) {
+    val coroutineScope = rememberCoroutineScope()
+    AndroidView(
+        modifier = Modifier
+            .fillMaxSize()
+            .padding(
+                horizontal = dimensionResource(id = R.dimen.home_item_horizontal_margin),
+            ),
+        factory = { context ->
+            WidgetSiteItemView(context)
```
Comment (gabrielluong) [inline fenix/app/src/main/java/org/mozilla/fenix/tabhistory/TabHistoryBottomSheetView.kt:141]: Sorry for the drive-by: Perhaps we should consider using [`FaviconListItem`](https://searchfox.org/mozilla-mobile/source/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/compose/list/ListItem.kt#98).

Comment (gabrielluong) [reply fenix/app/src/main/java/org/mozilla/fenix/tabhistory/TabHistoryBottomSheetView.kt:141]: You shouldn't have to worry about reproducing the box around the favicon. We made changes in bookmarks and history to remove the box around the favicon and I am guessing we never updated the Tab History view.

File: android-components/plugins/dependencies/src/main/java/DependenciesPlugin.kt (hunk 1 above)
Comment (iorgamgabriel) [reply android-components/plugins/dependencies/src/main/java/DependenciesPlugin.kt]: Done, reverted

File: fenix/app/src/main/java/org/mozilla/fenix/tabhistory/TabHistoryBottomSheetView.kt (hunk 2 above)
Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/tabhistory/TabHistoryBottomSheetView.kt]: Done

**Comment by gabrielluong** [comment]
Will get to looking at this soon this week.

File: fenix/app/src/main/java/org/mozilla/fenix/tabhistory/TabHistoryDialogFragment.kt [hunk 3]
```diff
@@ -1,72 +0,0 @@
...
-import kotlinx.coroutines.flow.distinctUntilChanged
-import kotlinx.coroutines.flow.mapNotNull
-import mozilla.components.browser.state.selector.findCustomTabOrSelectedTab
-import mozilla.components.lib.state.ext.flowScoped
-import mozilla.components.support.ktx.android.content.getColorFromAttr
-import org.mozilla.fenix.R
-import org.mozilla.fenix.databinding.FragmentTabHistoryDialogBinding
-import org.mozilla.fenix.ext.requireComponents
-
-class TabHistoryDialogFragment : BottomSheetDialogFragment() {
```
Comment (gabrielluong) [inline fenix/app/src/main/java/org/mozilla/fenix/tabhistory/TabHistoryDialogFragment.kt:24]: Sorry for taking so long to get to this review. I have some querying questions to better understand how you're approaching this. I was wondering if we could still keep the fragment navigation and implement this bottom sheet similar to:
https://searchfox.org/mozilla-mobile/source/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/shopping/ReviewQualityCheckFragment.kt
https://searchfox.org/mozilla-mobile/source/firefox-android/fenix/app/src/main/java/org/mozilla/fenix/translations/TranslationsDialogFragment.kt

Comment (iorgamgabriel) [reply fenix/app/src/main/java/org/mozilla/fenix/tabhistory/TabHistoryDialogFragment.kt:24]: I made this refactoring because it was a crash at navigation and we don't have steps to reproduce it. https://bugzilla.mozilla.org/show_bug.cgi?id=1804377
If I add navigation to this refactoring, it is almost useless because the crash still happens.

**Comment by rahulsainani** [comment]
@iorgamgabriel The bug marked is a nav crash, can this be separated so we fix the crash first and then refactor? Please let me know, I'd like this to move forward. Apologies, I couldn't check it before.

File: fenix/app/src/main/java/org/mozilla/fenix/tabhistory/TabHistoryController.kt [hunk 4]
```diff
@@ -4,22 +4,18 @@
...
 }
 
 class DefaultTabHistoryController(
-    private val navController: NavController,
     private val goToHistoryIndexUseCase: SessionUseCases.GoToHistoryIndexUseCase,
     private val customTabId: String? = null,
 ) : TabHistoryController {
 
     override fun handleGoToHistoryItem(item: TabHistoryItem) {
-        navController.navigateUp()
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/tabhistory/TabHistoryController.kt:21]: @iorgamgabriel @mcarare I find this suspicious. Have we tried simply changing this to `popBackStack`. 

Alternatively, have we tried passing a lamda similar that sets bottom sheet behavior to state hidden. Similar to examples @gabrielluong shared. 

From the docs

_Attempts to navigate up in the navigation hierarchy. Suitable for when the user presses the "Up" button marked with a left (or start)-facing arrow in the upper left (or starting) corner of the app UI.
The intended behavior of Up differs from Back when the user did not reach the current destination from the application's own task. e.g. if the user is viewing a document or link in the current app in an activity hosted on another app's task where the user clicked the link. In this case the current activity (determined by the context used to create this NavController) will be finished and the user will be taken to an appropriate destination in this app on its own task.
Returns:
true if navigation was successful, false otherwise_

**Comment by rahulsainani** [comment]
> Hi @rahulsainani We tried to fix the crash, but none of the fixes worked. This is the reason I did this refactoring: we don't have a fix for the crash.

I see, is it implying that bottom sheet and navigation don't work in tandem? I'm not convinced, I have asked a question about some suggestions above to fix this without the refactor.

Saying that, I think we should definitely use the UI rewrite to compose (keeping the fragment if the suggestion works) that you've done but that should be part of a different bug.

File: fenix/app/src/main/java/org/mozilla/fenix/tabhistory/TabHistoryController.kt (hunk 4 above)
Comment (mcarare) [reply fenix/app/src/main/java/org/mozilla/fenix/tabhistory/TabHistoryController.kt:21]: @rahulsainani You are welcome to try other paths to fix this. As long we don't have a STR I am not sure we can do more than a speculative fix.

**Comment by gabrielluong** [comment]
We have moved our firefox-android repo over to [mozilla-central](https://hg.mozilla.org/mozilla-central). Please feel free to reopen a new PR against the new repo. You will have to clone the new repository and reapply your changes. Thank you again, and sorry for the inconvenience.

For more information, please see https://github.com/mozilla-mobile/firefox-android/wiki#upcoming-migration-to-mozilla-central.
